### ISSUE TITLE
Add filtering for CRUD routes with query params

### ISSUE THREAD

**Issue by jpeter06** [mockoon/mockoon#1047]
Hello all,
working with routes and **CRUD**, would be nice to be able to apply **filters** to the responses.
Should be posible?
Just a beloved feature request

Thanks for this wonderfull mocker :-)

**Comment by 255kb** [comment]
I just updated the title a bit, to clarify 🙂

**Comment by 255kb** [comment]
@edtoken worked on a `filter` templating helper which could solve this issue. 
But now that I think about it, it would be good to also have built-in CRUD filtering with query params:
`/users?status=connected`
This is something json-server has and maybe what you had in mind @jpeter06 

I will change the title (again) to make this clear.

**Comment by HighLiuk** [comment]
@255kb hey there.
Why don't we allow search and filtering on CRUD routes?
search means I have a string and I want to look for entities containing that string in some of their fields.

Like this

```js
function search(input, query) {
  if (typeof input === 'boolean' || input === null) {
    return false;
  } else if (Array.isArray(input)) {
    return input.some((el) => search(el, query));
  } else if (typeof input === 'object') {
    return Object.values(input).some((value) => search(value, query));
  } else {
    return String(input).includes(query);
  }
}
```

```
GET /?search=hello
```

Filtering on the other hand is what you were discussing previously, namely "this parameter = this value". This may be accomplished without breaking changes on the parameter names with something like

```
GET /?filter[role]=admin&filter[email.status]=sent
```

How about this?

**Comment by 255kb** [comment]
@HighLiuk That's exactly the purpose of this issue 🙂
I'm not sure I see the difference between the search and the filtering. For me they are the same. 

But put everything behind a `filter` property is a good idea as it would avoid a breaking change.

**Comment by HighLiuk** [comment]
@255kb I'll try to write better the differences down 😁
Key differences between search and filter:

## search

- search is with includes and not ===, i.e. searching "h" filters out "world" but not "hello" if both are values of different rows
- search is field-agnostic: e.g. I can search my customer both by email, phone or name
- use case is when you have one search bar in the UI with general filtering

## filter

filter on the other hand is granular: 

- filters by exact match (e.g. drop-down)
- only look in the field of your choice (e.g. phone number but not birthday)
- can be combined with other filters due to its granular nature
- use case is when your UI has several filters on it

**Comment by 255kb** [comment]
ah! I get it.
I would say, in our case, I don't see why the filters could not allow for partial string matching. 
And yes, why not also add a search option. But it's more work. I would go for the filters first. 

If anyone wants to contribute, please let me know!

**Comment by 255kb** [comment]
Because json-server is not working in the same way. First it's a CLI, with external files, the state is persisted, etc. It's not exactly what we wanted for Mockoon: our data buckets are not persisting their state, on purpose, we support more than arrays of objects (you can use simple string, or arrays of booleans, for example). Also, now it's done like this and there is no real reason to ditch everything we did to replace with json-server and add lots of breaking changes.

**Comment by HighLiuk** [comment]
Got it. And indeed that is why I prefer Mockoon over it. DX is even better and that's great.
Agree with the fact that it should not be a dependency, but if not adding it as a dependency, why don't just get inspired by its implementation for filters etc.?
Like here: https://github.com/typicode/json-server/blob/main/src/server/router/plural.js#L51

**Comment by 255kb** [comment]
Yes, why not. We got inspired by their API, even if we didn't replicate everything.

**Comment by HighLiuk** [comment]
@255kb may I contribute?

**Comment by 255kb** [comment]
yes, @HighLiuk I will assign the issue to you and put it in progress on the roadmap. 
Let's catch up on discord (my username is 255kb) for the implementation details 🙂

**Comment by HighLiuk** [comment]
> yes, @HighLiuk I will assign the issue to you and put it in progress on the roadmap. Let's catch up on discord (my username is 255kb) for the implementation details 🙂

Sent you a request! 😉👍🏼

**Comment by 255kb** [comment]
📦 This is now released in [v6.1.0](https://mockoon.com/releases/6.1.0/).
